esp/bindings/http/platform/httpservice.cpp

```diff
@@ -974,3 +875,558 @@ int CEspHttpServer::onGet()
+        return;
+    }
+
+    ESPLOG(LogMin, "Authentication failed: call askUserLogin.");
+    if (sessionID)
+        askUserLogin(authReq.authBinding, *sessionID);
+    else
+        askUserLogin(authReq.authBinding, createHTTPSession(authReq, nullptr, nullptr));
+}
+
+void CEspHttpServer::askUserLogin(EspHttpBinding* authBinding, unsigned sessionID)
+{
+    VStringBuffer sessionIDStr("%u", sessionID);
+    addCookie(authBinding->getSessionIDCookieName(), sessionIDStr.str(), authBinding->getSessionTimeoutSeconds());
+    addCookie(SESSION_AUTH_COOKIE, "1", 0); //time out when browser is closed
+    m_response->redirect(*m_request, authBinding->getLoginURL());
+}
+
+void CEspHttpServer::send200OK()
+{
+    m_response->setContentType("text/html; charset=UTF-8");
+    m_response->setStatus(HTTP_STATUS_OK);
+    m_response->send();
+}
+
+unsigned CEspHttpServer::createHTTPSession(EspAuthRequest& authReq, IPropertyTree* domainSessions, const char* _loginURL)
+{
+    CDateTime now;
+    now.setNow();
+    time_t createTime = now.getSimple();
+
+    StringBuffer peer;
+    VStringBuffer idStr("%s_%ld", m_request->getPeer(peer).str(), createTime);
+    unsigned sessionID = hashc((unsigned char *)idStr.str(), idStr.length(), 0);
+    ESPLOG(LogMax, "New sessionID <%d> at <%ld> in createHTTPSession()", sessionID, createTime);
+
+    Owned<IPropertyTree> ptree = createPTree();
+    ptree->addProp(PropSessionNetworkAddress, peer.str());
+    ptree->addPropInt64(PropSessionID, sessionID);
+    ptree->addPropInt(PropSessionState, HTTPSS_new);
+    ptree->setPropInt64(PropSessionCreateTime, createTime);
+    ptree->setPropInt64(PropSessionLastAccessed, createTime);
+
+    if (!isEmptyString(_loginURL))
+        ptree->addProp(PropSessionLoginURL, _loginURL);
+    else
+    {
+        StringBuffer loginURL = authReq.httpPath;
+        if (authReq.requestParams && authReq.requestParams->hasProp("__querystring"))
+            loginURL.append("?").append(authReq.requestParams->queryProp("__querystring"));
+        if (!loginURL.isEmpty() && streq(loginURL.str(), "/WsSMC/"))
+            ptree->addProp(PropSessionLoginURL, "/");
+        else
+            ptree->addProp(PropSessionLoginURL, loginURL.str());
+    }
+
+    if (domainSessions)
+    {
+        domainSessions->addPropTree(PathSessionSession, LINK(ptree));
+        return sessionID;
+    }
+
+    Owned<IRemoteConnection> conn = querySDS().connect(authReq.authBinding->getDomainSessionSDSPath(), myProcessSession(), RTM_LOCK_WRITE, SESSION_SDS_LOCK_TIMEOUT);
+    if (!conn)
+        throw MakeStringException(-1, "Failed to connect SDS DomainSession.");
+
+    IPropertyTree* root = conn->queryRoot();
+    if (!root)
+        throw MakeStringException(-1, "Failed to get SDS DomainSession.");
+
+    root->addPropTree(PathSessionSession, LINK(ptree));
+    return sessionID;
+}
+
+IPropertyTree* CEspHttpServer::readAndCleanDomainSessions(EspHttpBinding* authBinding, void* _conn, time_t timeNow)
```

Comment:
Is that problematic?
And anyway you don't need the definition, you can forward declare.
e.g.:
```
interface IRemoteConnection;
class CEspHttpServer : implements IHttpServerService, public CInterface
{
....
    IPropertyTree* readAndCleanDomainSessions(EspHttpBinding* authBinding, IRemoteConnection* conn, time_t timeNow);
....
};

```